{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b881b469",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running on Apache Spark version 3.1.3\n",
      "SparkUI available at http://ibd-exome-qc-m.c.daly-ibd.internal:36127\n",
      "Welcome to\n",
      "     __  __     <>__\n",
      "    / /_/ /__  __/ /\n",
      "   / __  / _ `/ / /\n",
      "  /_/ /_/\\_,_/_/_/   version 0.2.99-57537fea08d4\n",
      "LOGGING: writing to /home/hail/hail-20230724-0148-0.2.99-57537fea08d4.log\n"
     ]
    }
   ],
   "source": [
    "import hail as hl\n",
    "# Initialize Hail \n",
    "hl.init(default_reference = 'GRCh38')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9cf19a60",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------\n",
      "Global fields:\n",
      "    None\n",
      "----------------------------------------\n",
      "Column fields:\n",
      "    's': str\n",
      "----------------------------------------\n",
      "Row fields:\n",
      "    'locus': locus<GRCh38>\n",
      "    'alleles': array<str>\n",
      "    'rsid': str\n",
      "    'info': struct {\n",
      "        AC_raw: array<int32>, \n",
      "        AN_raw: int32, \n",
      "        AF_raw: array<float64>, \n",
      "        nhomalt_raw: array<int32>, \n",
      "        AC: array<int32>, \n",
      "        AN: int32, \n",
      "        AF: array<float64>, \n",
      "        nhomalt: array<int32>\n",
      "    }\n",
      "    'a_index': int32\n",
      "    'was_split': bool\n",
      "    'nfe_ht': struct {\n",
      "        freq: array<struct {\n",
      "            AC: int32, \n",
      "            AF: float64, \n",
      "            AN: int32, \n",
      "            homozygote_count: int32\n",
      "        }>, \n",
      "        age_hist_het: array<struct {\n",
      "            bin_edges: array<float64>, \n",
      "            bin_freq: array<int64>, \n",
      "            n_smaller: int64, \n",
      "            n_larger: int64\n",
      "        }>, \n",
      "        age_hist_hom: array<struct {\n",
      "            bin_edges: array<float64>, \n",
      "            bin_freq: array<int64>, \n",
      "            n_smaller: int64, \n",
      "            n_larger: int64\n",
      "        }>, \n",
      "        popmax: array<struct {\n",
      "            AC: int32, \n",
      "            AF: float64, \n",
      "            AN: int32, \n",
      "            homozygote_count: int32, \n",
      "            pop: str\n",
      "        }>, \n",
      "        faf: array<struct {\n",
      "            meta: dict<str, str>, \n",
      "            faf95: float64, \n",
      "            faf99: float64\n",
      "        }>, \n",
      "        lcr: bool, \n",
      "        decoy: bool, \n",
      "        segdup: bool, \n",
      "        nonpar: bool, \n",
      "        variant_type: str, \n",
      "        allele_type: str, \n",
      "        n_alt_alleles: int32, \n",
      "        was_mixed: bool, \n",
      "        has_star: bool, \n",
      "        qd: float64, \n",
      "        pab_max: float64, \n",
      "        info_MQRankSum: float64, \n",
      "        info_SOR: float64, \n",
      "        info_InbreedingCoeff: float64, \n",
      "        info_ReadPosRankSum: float64, \n",
      "        info_FS: float64, \n",
      "        info_QD: float64, \n",
      "        info_MQ: float64, \n",
      "        info_DP: int32, \n",
      "        transmitted_singleton: bool, \n",
      "        fail_hard_filters: bool, \n",
      "        info_POSITIVE_TRAIN_SITE: bool, \n",
      "        info_NEGATIVE_TRAIN_SITE: bool, \n",
      "        omni: bool, \n",
      "        mills: bool, \n",
      "        n_nonref: int32, \n",
      "        tp: bool, \n",
      "        rf_train: bool, \n",
      "        rf_label: str, \n",
      "        rf_probability: float64, \n",
      "        singleton: bool, \n",
      "        was_split: bool, \n",
      "        score: float64, \n",
      "        rank: int64, \n",
      "        singleton_rank: int64, \n",
      "        biallelic_rank: int64, \n",
      "        adj_biallelic_singleton_rank: int64, \n",
      "        adj_rank: int64, \n",
      "        adj_biallelic_rank: int64, \n",
      "        adj_singleton_rank: int64, \n",
      "        biallelic_singleton_rank: int64, \n",
      "        filters: set<str>, \n",
      "        gq_hist_alt: struct {\n",
      "            bin_edges: array<float64>, \n",
      "            bin_freq: array<int64>, \n",
      "            n_smaller: int64, \n",
      "            n_larger: int64\n",
      "        }, \n",
      "        gq_hist_all: struct {\n",
      "            bin_edges: array<float64>, \n",
      "            bin_freq: array<int64>, \n",
      "            n_smaller: int64, \n",
      "            n_larger: int64\n",
      "        }, \n",
      "        dp_hist_alt: struct {\n",
      "            bin_edges: array<float64>, \n",
      "            bin_freq: array<int64>, \n",
      "            n_smaller: int64, \n",
      "            n_larger: int64\n",
      "        }, \n",
      "        dp_hist_all: struct {\n",
      "            bin_edges: array<float64>, \n",
      "            bin_freq: array<int64>, \n",
      "            n_smaller: int64, \n",
      "            n_larger: int64\n",
      "        }, \n",
      "        ab_hist_alt: struct {\n",
      "            bin_edges: array<float64>, \n",
      "            bin_freq: array<int64>, \n",
      "            n_smaller: int64, \n",
      "            n_larger: int64\n",
      "        }, \n",
      "        qual: float64, \n",
      "        vep: struct {\n",
      "            assembly_name: str, \n",
      "            allele_string: str, \n",
      "            ancestral: str, \n",
      "            colocated_variants: array<struct {\n",
      "                aa_allele: str, \n",
      "                aa_maf: float64, \n",
      "                afr_allele: str, \n",
      "                afr_maf: float64, \n",
      "                allele_string: str, \n",
      "                amr_allele: str, \n",
      "                amr_maf: float64, \n",
      "                clin_sig: array<str>, \n",
      "                end: int32, \n",
      "                eas_allele: str, \n",
      "                eas_maf: float64, \n",
      "                ea_allele: str, \n",
      "                ea_maf: float64, \n",
      "                eur_allele: str, \n",
      "                eur_maf: float64, \n",
      "                exac_adj_allele: str, \n",
      "                exac_adj_maf: float64, \n",
      "                exac_allele: str, \n",
      "                exac_afr_allele: str, \n",
      "                exac_afr_maf: float64, \n",
      "                exac_amr_allele: str, \n",
      "                exac_amr_maf: float64, \n",
      "                exac_eas_allele: str, \n",
      "                exac_eas_maf: float64, \n",
      "                exac_fin_allele: str, \n",
      "                exac_fin_maf: float64, \n",
      "                exac_maf: float64, \n",
      "                exac_nfe_allele: str, \n",
      "                exac_nfe_maf: float64, \n",
      "                exac_oth_allele: str, \n",
      "                exac_oth_maf: float64, \n",
      "                exac_sas_allele: str, \n",
      "                exac_sas_maf: float64, \n",
      "                id: str, \n",
      "                minor_allele: str, \n",
      "                minor_allele_freq: float64, \n",
      "                phenotype_or_disease: int32, \n",
      "                pubmed: array<int32>, \n",
      "                sas_allele: str, \n",
      "                sas_maf: float64, \n",
      "                somatic: int32, \n",
      "                start: int32, \n",
      "                strand: int32\n",
      "            }>, \n",
      "            context: str, \n",
      "            end: int32, \n",
      "            id: str, \n",
      "            input: str, \n",
      "            intergenic_consequences: array<struct {\n",
      "                allele_num: int32, \n",
      "                consequence_terms: array<str>, \n",
      "                impact: str, \n",
      "                minimised: int32, \n",
      "                variant_allele: str\n",
      "            }>, \n",
      "            most_severe_consequence: str, \n",
      "            motif_feature_consequences: array<struct {\n",
      "                allele_num: int32, \n",
      "                consequence_terms: array<str>, \n",
      "                high_inf_pos: str, \n",
      "                impact: str, \n",
      "                minimised: int32, \n",
      "                motif_feature_id: str, \n",
      "                motif_name: str, \n",
      "                motif_pos: int32, \n",
      "                motif_score_change: float64, \n",
      "                strand: int32, \n",
      "                variant_allele: str\n",
      "            }>, \n",
      "            regulatory_feature_consequences: array<struct {\n",
      "                allele_num: int32, \n",
      "                biotype: str, \n",
      "                consequence_terms: array<str>, \n",
      "                impact: str, \n",
      "                minimised: int32, \n",
      "                regulatory_feature_id: str, \n",
      "                variant_allele: str\n",
      "            }>, \n",
      "            seq_region_name: str, \n",
      "            start: int32, \n",
      "            strand: int32, \n",
      "            transcript_consequences: array<struct {\n",
      "                allele_num: int32, \n",
      "                amino_acids: str, \n",
      "                biotype: str, \n",
      "                canonical: int32, \n",
      "                ccds: str, \n",
      "                cdna_start: int32, \n",
      "                cdna_end: int32, \n",
      "                cds_end: int32, \n",
      "                cds_start: int32, \n",
      "                codons: str, \n",
      "                consequence_terms: array<str>, \n",
      "                distance: int32, \n",
      "                domains: array<struct {\n",
      "                    db: str, \n",
      "                    name: str\n",
      "                }>, \n",
      "                exon: str, \n",
      "                gene_id: str, \n",
      "                gene_pheno: int32, \n",
      "                gene_symbol: str, \n",
      "                gene_symbol_source: str, \n",
      "                hgnc_id: str, \n",
      "                hgvsc: str, \n",
      "                hgvsp: str, \n",
      "                hgvs_offset: int32, \n",
      "                impact: str, \n",
      "                intron: str, \n",
      "                lof: str, \n",
      "                lof_flags: str, \n",
      "                lof_filter: str, \n",
      "                lof_info: str, \n",
      "                minimised: int32, \n",
      "                polyphen_prediction: str, \n",
      "                polyphen_score: float64, \n",
      "                protein_end: int32, \n",
      "                protein_start: int32, \n",
      "                protein_id: str, \n",
      "                sift_prediction: str, \n",
      "                sift_score: float64, \n",
      "                strand: int32, \n",
      "                swissprot: str, \n",
      "                transcript_id: str, \n",
      "                trembl: str, \n",
      "                uniparc: str, \n",
      "                variant_allele: str, \n",
      "                most_severe_consequence: str\n",
      "            }>, \n",
      "            variant_class: str, \n",
      "            worst_consequence_term: str, \n",
      "            worst_csq_by_gene: array<struct {\n",
      "                allele_num: int32, \n",
      "                amino_acids: str, \n",
      "                biotype: str, \n",
      "                canonical: int32, \n",
      "                ccds: str, \n",
      "                cdna_start: int32, \n",
      "                cdna_end: int32, \n",
      "                cds_end: int32, \n",
      "                cds_start: int32, \n",
      "                codons: str, \n",
      "                consequence_terms: array<str>, \n",
      "                distance: int32, \n",
      "                domains: array<struct {\n",
      "                    db: str, \n",
      "                    name: str\n",
      "                }>, \n",
      "                exon: str, \n",
      "                gene_id: str, \n",
      "                gene_pheno: int32, \n",
      "                gene_symbol: str, \n",
      "                gene_symbol_source: str, \n",
      "                hgnc_id: str, \n",
      "                hgvsc: str, \n",
      "                hgvsp: str, \n",
      "                hgvs_offset: int32, \n",
      "                impact: str, \n",
      "                intron: str, \n",
      "                lof: str, \n",
      "                lof_flags: str, \n",
      "                lof_filter: str, \n",
      "                lof_info: str, \n",
      "                minimised: int32, \n",
      "                polyphen_prediction: str, \n",
      "                polyphen_score: float64, \n",
      "                protein_end: int32, \n",
      "                protein_start: int32, \n",
      "                protein_id: str, \n",
      "                sift_prediction: str, \n",
      "                sift_score: float64, \n",
      "                strand: int32, \n",
      "                swissprot: str, \n",
      "                transcript_id: str, \n",
      "                trembl: str, \n",
      "                uniparc: str, \n",
      "                variant_allele: str, \n",
      "                most_severe_consequence: str, \n",
      "                csq_score: float64\n",
      "            }>, \n",
      "            any_lof: bool, \n",
      "            gene_with_most_severe_csq: array<str>, \n",
      "            ensg_with_most_severe_csq: array<str>\n",
      "        }, \n",
      "        allele_info: struct {\n",
      "            BaseQRankSum: float64, \n",
      "            ClippingRankSum: float64, \n",
      "            DB: bool, \n",
      "            DP: int32, \n",
      "            DS: bool, \n",
      "            END: int32, \n",
      "            FS: float64, \n",
      "            HaplotypeScore: float64, \n",
      "            InbreedingCoeff: float64, \n",
      "            MQ: float64, \n",
      "            MQRankSum: float64, \n",
      "            NEGATIVE_TRAIN_SITE: bool, \n",
      "            POSITIVE_TRAIN_SITE: bool, \n",
      "            QD: float64, \n",
      "            ReadPosRankSum: float64, \n",
      "            SOR: float64, \n",
      "            VQSLOD: float64, \n",
      "            culprit: str\n",
      "        }, \n",
      "        rsid: str, \n",
      "        fin: struct {\n",
      "            AC: int32, \n",
      "            AF: float64, \n",
      "            AN: int32, \n",
      "            homozygote_count: int32\n",
      "        }, \n",
      "        nfe: struct {\n",
      "            AC: int32, \n",
      "            AF: float64, \n",
      "            AN: int32, \n",
      "            homozygote_count: int32\n",
      "        }, \n",
      "        nfe_ds: struct {\n",
      "            AC: int32, \n",
      "            AF: float64, \n",
      "            AN: int32, \n",
      "            homozygote_count: int32\n",
      "        }, \n",
      "        nfsee: struct {\n",
      "            AC: int32, \n",
      "            AN: int32, \n",
      "            homozygote_count: int32, \n",
      "            AF: float32\n",
      "        }, \n",
      "        consequence: str, \n",
      "        freq_bin_fin: float64, \n",
      "        freq_bin_nfe: float64, \n",
      "        freq_bin_nfe_ds: float64, \n",
      "        freq_bin_nfsee: float64, \n",
      "        enrichment: float64, \n",
      "        enrichment_ds: float64, \n",
      "        enrichment_nfsee: float64, \n",
      "        enrichment_pseudo: float64, \n",
      "        enrichment_pseudo_ds: float64, \n",
      "        enrichment_pseudo_nfsee: float64, \n",
      "        fet: struct {\n",
      "            p_value: float64, \n",
      "            odds_ratio: float64\n",
      "        }, \n",
      "        fet_ds: struct {\n",
      "            p_value: float64, \n",
      "            odds_ratio: float64\n",
      "        }, \n",
      "        fet_nfsee: struct {\n",
      "            p_value: float64, \n",
      "            odds_ratio: float64\n",
      "        }, \n",
      "        enrichment_level: float64, \n",
      "        enrichment_level_ds: float64, \n",
      "        enrichment_level_nfsee: float64, \n",
      "        enrichment_level_pseudo: float64, \n",
      "        enrichment_level_pseudo_ds: float64, \n",
      "        enrichment_level_pseudo_nfsee: float64, \n",
      "        enrichment_trunc: float64, \n",
      "        enrichment_trunc_ds: float64, \n",
      "        enrichment_trunc_nfsee: float64, \n",
      "        enrichment_trunc_pseudo: float64, \n",
      "        enrichment_trunc_pseudo_ds: float64, \n",
      "        enrichment_trunc_pseudo_nfsee: float64, \n",
      "        new_locus: locus<GRCh38>\n",
      "    }\n",
      "    'vep': struct {\n",
      "        assembly_name: str, \n",
      "        allele_string: str, \n",
      "        ancestral: str, \n",
      "        colocated_variants: array<struct {\n",
      "            aa_allele: str, \n",
      "            aa_maf: float64, \n",
      "            afr_allele: str, \n",
      "            afr_maf: float64, \n",
      "            allele_string: str, \n",
      "            amr_allele: str, \n",
      "            amr_maf: float64, \n",
      "            clin_sig: array<str>, \n",
      "            end: int32, \n",
      "            eas_allele: str, \n",
      "            eas_maf: float64, \n",
      "            ea_allele: str, \n",
      "            ea_maf: float64, \n",
      "            eur_allele: str, \n",
      "            eur_maf: float64, \n",
      "            exac_adj_allele: str, \n",
      "            exac_adj_maf: float64, \n",
      "            exac_allele: str, \n",
      "            exac_afr_allele: str, \n",
      "            exac_afr_maf: float64, \n",
      "            exac_amr_allele: str, \n",
      "            exac_amr_maf: float64, \n",
      "            exac_eas_allele: str, \n",
      "            exac_eas_maf: float64, \n",
      "            exac_fin_allele: str, \n",
      "            exac_fin_maf: float64, \n",
      "            exac_maf: float64, \n",
      "            exac_nfe_allele: str, \n",
      "            exac_nfe_maf: float64, \n",
      "            exac_oth_allele: str, \n",
      "            exac_oth_maf: float64, \n",
      "            exac_sas_allele: str, \n",
      "            exac_sas_maf: float64, \n",
      "            id: str, \n",
      "            minor_allele: str, \n",
      "            minor_allele_freq: float64, \n",
      "            phenotype_or_disease: int32, \n",
      "            pubmed: array<int32>, \n",
      "            sas_allele: str, \n",
      "            sas_maf: float64, \n",
      "            somatic: int32, \n",
      "            start: int32, \n",
      "            strand: int32\n",
      "        }>, \n",
      "        context: str, \n",
      "        end: int32, \n",
      "        id: str, \n",
      "        input: str, \n",
      "        intergenic_consequences: array<struct {\n",
      "            allele_num: int32, \n",
      "            consequence_terms: array<str>, \n",
      "            impact: str, \n",
      "            minimised: int32, \n",
      "            variant_allele: str\n",
      "        }>, \n",
      "        most_severe_consequence: str, \n",
      "        motif_feature_consequences: array<struct {\n",
      "            allele_num: int32, \n",
      "            consequence_terms: array<str>, \n",
      "            high_inf_pos: str, \n",
      "            impact: str, \n",
      "            minimised: int32, \n",
      "            motif_feature_id: str, \n",
      "            motif_name: str, \n",
      "            motif_pos: int32, \n",
      "            motif_score_change: float64, \n",
      "            strand: int32, \n",
      "            variant_allele: str\n",
      "        }>, \n",
      "        regulatory_feature_consequences: array<struct {\n",
      "            allele_num: int32, \n",
      "            biotype: str, \n",
      "            consequence_terms: array<str>, \n",
      "            impact: str, \n",
      "            minimised: int32, \n",
      "            regulatory_feature_id: str, \n",
      "            variant_allele: str\n",
      "        }>, \n",
      "        seq_region_name: str, \n",
      "        start: int32, \n",
      "        strand: int32, \n",
      "        transcript_consequences: array<struct {\n",
      "            allele_num: int32, \n",
      "            amino_acids: str, \n",
      "            appris: str, \n",
      "            biotype: str, \n",
      "            canonical: int32, \n",
      "            ccds: str, \n",
      "            cdna_start: int32, \n",
      "            cdna_end: int32, \n",
      "            cds_end: int32, \n",
      "            cds_start: int32, \n",
      "            codons: str, \n",
      "            consequence_terms: array<str>, \n",
      "            distance: int32, \n",
      "            domains: array<struct {\n",
      "                db: str, \n",
      "                name: str\n",
      "            }>, \n",
      "            exon: str, \n",
      "            gene_id: str, \n",
      "            gene_pheno: int32, \n",
      "            gene_symbol: str, \n",
      "            gene_symbol_source: str, \n",
      "            hgnc_id: str, \n",
      "            hgvsc: str, \n",
      "            hgvsp: str, \n",
      "            hgvs_offset: int32, \n",
      "            impact: str, \n",
      "            intron: str, \n",
      "            lof: str, \n",
      "            lof_flags: str, \n",
      "            lof_filter: str, \n",
      "            lof_info: str, \n",
      "            minimised: int32, \n",
      "            polyphen_prediction: str, \n",
      "            polyphen_score: float64, \n",
      "            protein_end: int32, \n",
      "            protein_start: int32, \n",
      "            protein_id: str, \n",
      "            sift_prediction: str, \n",
      "            sift_score: float64, \n",
      "            strand: int32, \n",
      "            swissprot: str, \n",
      "            transcript_id: str, \n",
      "            trembl: str, \n",
      "            tsl: int32, \n",
      "            uniparc: str, \n",
      "            variant_allele: str\n",
      "        }>, \n",
      "        variant_class: str\n",
      "    }\n",
      "----------------------------------------\n",
      "Entry fields:\n",
      "    'GT': call\n",
      "    'DP': int32\n",
      "    'GQ': int32\n",
      "    'RGQ': int32\n",
      "    'gvcf_info': struct {\n",
      "        AC: array<int32>, \n",
      "        AF: array<float64>, \n",
      "        AN: int32, \n",
      "        AS_BaseQRankSum: array<float64>, \n",
      "        AS_FS: array<float64>, \n",
      "        AS_InbreedingCoeff: array<float64>, \n",
      "        AS_MQ: array<float64>, \n",
      "        AS_MQRankSum: array<float64>, \n",
      "        AS_QD: array<float64>, \n",
      "        AS_QUALapprox: array<int32>, \n",
      "        AS_RAW_BaseQRankSum: str, \n",
      "        AS_RAW_MQ: array<float64>, \n",
      "        AS_RAW_MQRankSum: array<tuple (\n",
      "            float64, \n",
      "            int32\n",
      "        )>, \n",
      "        AS_RAW_ReadPosRankSum: array<tuple (\n",
      "            float64, \n",
      "            int32\n",
      "        )>, \n",
      "        AS_ReadPosRankSum: array<float64>, \n",
      "        AS_SB_TABLE: array<array<int32>>, \n",
      "        AS_SOR: array<float64>, \n",
      "        AS_VarDP: array<int32>, \n",
      "        BaseQRankSum: float64, \n",
      "        ExcessHet: float64, \n",
      "        FS: float64, \n",
      "        InbreedingCoeff: float64, \n",
      "        MQ: float64, \n",
      "        MQRankSum: float64, \n",
      "        MQ_DP: int32, \n",
      "        QD: float64, \n",
      "        QUALapprox: int32, \n",
      "        RAW_GT_COUNT: array<int32>, \n",
      "        RAW_MQandDP: array<int32>, \n",
      "        ReadPosRankSum: float64, \n",
      "        SOR: float64, \n",
      "        VarDP: int32\n",
      "    }\n",
      "    'MIN_DP': int32\n",
      "    'PID': str\n",
      "    'SB': array<int32>\n",
      "    'PGT': call\n",
      "    'AD': array<int32>\n",
      "    'PL': array<int32>\n",
      "----------------------------------------\n",
      "Column key: ['s']\n",
      "Row key: ['locus', 'alleles']\n",
      "----------------------------------------\n"
     ]
    }
   ],
   "source": [
    "mt = hl.read_matrix_table(\"gs://ibd-exomes-gnomad-subset/QC_round2/2.variant_narrowing/variant_narrowed.mt\")\n",
    "mt.describe()\n",
    "# mt.show(20)\n",
    "# mt.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "44ba229b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #gnomad quality filtering - read in tables######################\n",
    "# print(\"Reading in gNOMAD tables...\")\n",
    "\n",
    "# gnomad_exomes_v2_ht = hl.read_table(\"gs://gcp-public-data--gnomad/release/2.1.1/liftover_grch38/ht/exomes/gnomad.exomes.r2.1.1.sites.liftover_grch38.ht\")\n",
    "# gnomad_genomes_v2_ht = hl.read_table(\"gs://gcp-public-data--gnomad/release/2.1.1/liftover_grch38/ht/genomes/gnomad.genomes.r2.1.1.sites.liftover_grch38.ht\")\n",
    "# # print(\"Annotating gNOMAD exomes and genomes fields...\")\n",
    "# mt = mt.annotate_rows(gnomad_exomes = gnomad_exomes_v2_ht[mt.row_key])\n",
    "# mt = mt.annotate_rows(gnomad_genomes = gnomad_genomes_v2_ht[mt.row_key])\n",
    "\n",
    "# mt = mt.annotate_rows(GE_Filter = (mt.gnomad_exomes.filters.length() > 0) | mt.gnomad_exomes.filters.contains(\"RF\"))\n",
    "# mt = mt.annotate_rows(GG_Filter = (mt.gnomad_genomes.filters.length() > 0) | mt.gnomad_genomes.filters.contains(\"RF\"))\n",
    "\n",
    "# # print(\"Filtering for PASSing variants...\")\n",
    "# mt = mt.filter_rows(mt.GG_Filter | mt.GE_Filter, keep=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0b8b169b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(25105651, 198597)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Set GQ to missing if low\n",
    "mt = mt.annotate_entries(GT = hl.or_missing(mt.GQ > 20, mt.GT))\n",
    "mt = hl.variant_qc(mt, name='variant_qc')\n",
    "mt.count()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "854ef76c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-07-24 05:23:08 Hail: INFO: merging 47961 files totalling 52.4M...\n",
      "2023-07-24 05:24:05 Hail: INFO: while writing:\n",
      "    gs://ibd-exomes-gnomad-subset/QC_round2/low_call_rate_vars.tsv.gz\n",
      "  merge time: 56.732s\n"
     ]
    }
   ],
   "source": [
    "# mt.variant_qc.call_rate.show(500)\n",
    "mt.variant_qc.call_rate.export('gs://ibd-exomes-gnomad-subset/QC_round2/low_call_rate_vars.tsv.gz')\n",
    "# cr_hist = mt.aggregate_entries(hl.expr.aggregators.hist(mt.variant_qc.call_rate, 0, 1, 30))\n",
    "# p = hl.plot.histogram(cr_hist)\n",
    "# show(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3794c5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## ---------------------------------------------- variant QC --------------------------------------------------\n",
    "# Set GQ to missing if low\n",
    "mt = mt.annotate_entries(GT = hl.or_missing(mt.GQ > 20, mt.GT))\n",
    "mt = hl.variant_qc(mt, name='variant_qc')\n",
    "mt = mt.filter_rows(mt.variant_qc.call_rate > 0.95, keep = True)\n",
    "# mt = mt.filter_rows(mt.variant_qc.AC[0] > 0, keep = True)\n",
    "mt = mt.filter_rows(mt.variant_qc.gq_stats.mean > 10, keep = True)\n",
    "\n",
    "## ---------------------------------------------- sample QC --------------------------------------------------\n",
    "mt = hl.sample_qc(mt, name='sample_qc')\n",
    "mt = mt.filter_cols(mt.sample_qc.n_singleton < 500, keep = True)\n",
    "mt = mt.filter_cols(mt.sample_qc.gq_stats.mean > 30, keep = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "341b8706",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub message rate exceeded.\n",
      "The notebook server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--NotebookApp.iopub_msg_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)\n",
      "NotebookApp.rate_limit_window=3.0 (secs)\n",
      "\n",
      "IOPub message rate exceeded.\n",
      "The notebook server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--NotebookApp.iopub_msg_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)\n",
      "NotebookApp.rate_limit_window=3.0 (secs)\n",
      "\n",
      "IOPub message rate exceeded.\n",
      "The notebook server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--NotebookApp.iopub_msg_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "NotebookApp.iopub_msg_rate_limit=1000.0 (msgs/sec)\n",
      "NotebookApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "mt = mt.write(\"gs://ibd-exomes-gnomad-subset/QC_round2/3.variant+sampleQC/variant_sample_QCed.mt\", overwrite=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b138ad74",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}